src/SubHask/Algebra/Accelerate/Vector.hs

```diff
@@ -43,183 +37,198 @@ import Unsafe.Coerce
   where
+    --dim :: ACCVector b (n::Nat) r -> Index(A.Exp Int)
     {-# INLINE dim #-}
     dim (ACCVector v) = A.size v
 
 
-
+-- Could not deduce (r ~ Elem r)
 instance
-    ( P.Num (A.Exp r)
-    , Monoid r
+    ( Monoid r
     , ValidLogic r
     , ValidACCVector b n r
-    , IsScalar r
+    --, IsScalar r
+    , KnownNat n
     , FreeModule r
-    ) => IxContainer (ACCVector b (n::Symbol) r)
+    ) => IxContainer (ACCVector b (n::Nat) r)
         where
 
     {-# INLINE (!) #-}
-    (!) (ACCVector v) i =  v A.! (A.index1 (A.lift i))
+    (!) (ACCVector v) i =  A.the (v A.! A.index1 i)
 
-    --Couldn't match type ‘A.Exp Bool’ with ‘Bool’
     {-# INLINABLE imap #-}
     imap f (ACCVector v) = let
-      shp = A.shape v
-      idxs = A.generate shp P.id
-      mpd = A.zipWith f idxs v :: f (A.Exp r) -> f (A.Exp r) -> f (A.Exp r)
+      mpd = A.imap (\x i -> f i x) v
       in ACCVector mpd
 
-    type ValidElem (ACCVector b n r) e = (ClassicalLogic e, IsScalar e, FiniteModule e, ValidACCVector b n e)
+    type ValidElem (ACCVector b n r) e = (IsScalar e, FiniteModule e, ValidACCVector b n e)
 
-instance (Eq r, Monoid r, ValidACCVector b n r) => Eq_ (ACCVector b (n::Symbol) r) where
+instance (A.Eq r, KnownNat n, Eq r, Monoid r, ValidACCVector b n r) => Eq_ (ACCVector b (n::Nat) r) where
+    --(==) :: ACCVector b n r -> ACCVector b n r -> A.Acc (A.Scalar Bool)
     {-# INLINE (==) #-}
     (ACCVector v2) == (ACCVector v1) = let
-      l = (A.lift v1) A.==* (A.lift v2)
-      in l
+      l = A.zipWith (A.==*) v1 v2
+      ele = l A.! A.index1 (A.constant 0)
+      bl = A.all (A.&&* ele) l
+      in A.the bl
 
 instance
     ( ValidACCVector b n r
-    , P.Num (A.Exp r)
     , ExpField r
-    , Normed r
+    --, Normed r
     , Ord_ r
-    , Logic r~ Bool
     , IsScalar r
     , VectorSpace r
-    ) => Metric (ACCVector b (n::Symbol) r)
+    , KnownNat n
+    ) => Metric (ACCVector b (n::Nat) r)
 
         where
     {-# INLINE[2] distance #-}
     distance (ACCVector v1) (ACCVector v2) = {-# SCC distance_ACCVector #-}let
       dmag = A.zipWith (P.-) v1 v2
       dsq = A.zipWith (P.*) dmag dmag
       drt = A.sqrt (A.sum dsq)
-      in A.lift (A.the drt)
+      in A.the drt
 
-instance (VectorSpace r, ValidACCVector b n r, IsScalar r, ExpField r) => Normed (ACCVector b (n::Symbol) r) where
+instance (P.Floating (A.Acc (A.Array A.DIM0 r)), KnownNat n, VectorSpace r, ValidACCVector b n r, IsScalar r, ExpField r) => Normed (ACCVector b (n::Nat) r) where
     {-# INLINE size #-}
+    --Could not deduce (r ~ A.Exp r)
+    size :: ACCVector b (n::Nat) r -> A.Exp r
     size (ACCVector v1) = let
-      sq = A.zipWith (P.*) v1 v1
+      sq = A.zipWith (P.*) v1 v1 :: A.Acc (A.Array A.DIM1 r)
       s = A.fold (P.+) (A.constant 0.0) sq
-      in A.the (A.sqrt s)
+      srt = A.sqrt (s::A.Acc (A.Array A.DIM0 r))
+      in A.the srt :: A.Exp r
+
+
+-- -- Couldn't match type ‘A.Acc (Scalar Bool)’ with ‘Bool’
 
 instance
```

Comment:
`Banach`, `Hilbert`, `Normed`, and `TensorAlgebra` complain that about being unable to match `A.Exp Bool` with `Bool`.  The only `Bool` type I've added is in the `Logic` instance in line 43, so I'm not totally sure where the conflict is unless something in `SubHask.Algebra` I don't understand is requiring a `Bool` from something.  

I'm beginning to suspect it has something to do with the crazy amount of stuff in `ValidACCVector`.

Full error (which is the same for each of those instances):

```
Users/timpierson/subhask/src/SubHask/Algebra/Accelerate/Vector.hs: 180, 5

Couldn't match type ‘A.Exp Bool’ with ‘Bool’
In the instance declaration for ‘Metric (ACCVector b n r)’

/Users/timpierson/subhask/src/SubHask/Algebra/Accelerate/Vector.hs: 180, 5

Couldn't match type ‘A.Exp Bool’ with ‘Bool’
Inaccessible code in the instance declaration
In the instance declaration for ‘Metric (ACCVector b n r)’ 
```